### PR TITLE
[Enhancement] ranking window function support limit push down

### LINKED ISSUE #7739: [Enhancement] ranking window function with topn optimization
## Background

As for ranking window functions, we have already done some optimizations in the case of ranking predicate, refer to #5885 for details. And we can perform the similar optimization when ranking window function is followed by a group by + limit, the example sql is as follows: 

```sql
select * from (
    select *, rank() over (partition by v2 order by v3) as rk from t0
) sub_t0
order by rk limit 5;
```

The corresponding logical plan turns out to be:

```
TOPN (order by rk, limit 5)
            ↑
WINDOW_FUNCTION ( partition by v2 order by v3 )
            ↑
SORT ( order by v2, v3 )
            ↑
EXCHANGE ( shuffle by v2 )
            ↑
SCAN
```

**There are two main flaws of this plan:**

1. `EXCHANGE` Node needs to shuffle full data
2. `SORT` Node perform a machine level full sort

## Design

**Notice that there exists the order by limit `order by rk limit`, we can add a `PARTITION_TOPN` node before exchange node in order to filter data, the expected optimized plan is as follows:**

```
TOPN (order by rk, limit 5)
            ↑
WINDOW_FUNCTION ( partition by v2 order by v3 )
            ↑
SORT ( order by v2, v3 )
            ↑
EXCHANGE ( shuffle by v2 )
            ↑
PARTITION_TOPN ( partition by v2, sort by v3, limit by rank )
            ↑
SCAN
```

**It should be emphasized that the output of the `PARTITION_TOPN` is still remained unsorted, the purpose of this operator is to filter data**

If you are interested in `PARTITION_TOPN`, referring to #5885 for details

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #7739

## Enhancement

```sql
-- Q1
select * from (
    select *, rank() over (order by v3) as rk from t0
) sub_t0
order by rk limit 5;

-- Q2
select * from (
    select *, rank() over (partition by v2 order by v3) as rk from t0
) sub_t0
order by rk limit 2, 5;
```

**The optimization is similar to the previous work, by introducing a component `partitioner` to filter data. And I won't go into details about the implementation of `partitioner` here, you can refer to #5885 for design details**

**We will add a transform rule named `PushDownLimitRankingWindowRule` to do the optimitation**

* For query Q1, we can change the `sort operator(order by v3)` below the `window operator` to `topn operator(order by v3, limit 5)`
* For query Q2, we can add an extra `partition topn operator(order by v2,v3, limit 7)` before `window operator`

**And, it should be emphasized that the above optimization can be applied if and only if the following conditions are met**

* no other window function share the same order group with the ranking window function, because we cannot eliminate the full sort, so there is no need to do this optimization( may be lead to negative gain )
    ```sql
    -- Counter example
    select * from (
        select *, 
        rank() over (order by v3) as rk,
        sum(v1) over (order by v3) as sum_v1
        from t0
    ) sub_t0
    order by rk limit 5;
    ```
* the ranking column must be the only one or the first column of the order by columns
    ```sql
    -- Counter example
    select * from (
        select *, rank() over (order by v3) as rk from t0
    ) sub_t0
    order by v1, rk limit 5;
    ```
* the ranking order by column must be `asc`
    ```sql
    -- Counter example
    select * from (
        select *, rank() over (order by v3) as rk from t0
    ) sub_t0
    order by rk desc limit 5;
    ```

## Test

1fe/3be(64c 128g)
tpcds-100g

The following sql is modified from Q-67 by changing `where rk < 100` to `order by rk,xxx limit 100`

```sql
select  *
from (select i_category
            ,i_class
            ,i_brand
            ,i_product_name
            ,d_year
            ,d_qoy
            ,d_moy
            ,s_store_id
            ,sumsales
            ,rank() over (partition by i_category order by sumsales desc) rk
      from (select i_category
                  ,i_class
                  ,i_brand
                  ,i_product_name
                  ,d_year
                  ,d_qoy
                  ,d_moy
                  ,s_store_id
                  ,sum(coalesce(ss_sales_price*ss_quantity,0)) sumsales
            from store_sales
                ,date_dim
                ,store
                ,item
       where  ss_sold_date_sk=d_date_sk
          and ss_item_sk=i_item_sk
          and ss_store_sk = s_store_sk
          and d_month_seq between 1200 and 1200+11
       group by  rollup(i_category, i_class, i_brand, i_product_name, d_year, d_qoy, d_moy,s_store_id))dw1) dw2
order by rk, i_category
        ,i_class
        ,i_brand
        ,i_product_name
        ,d_year
        ,d_qoy
        ,d_moy
        ,s_store_id
        ,sumsales
        ,rk
limit 100;
```

| Before | After |
|:--|:--|
| 20s | 7.8s |

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 63 / 66 (95.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java|58|61|95.08%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateRankingWindowRule.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                .setSortPhase(sortPhase)
+                .build(), grandChildExpr.getInputs());
+
+        OptExpression newWindowOptExp =
+                OptExpression.create(new LogicalWindowOperator.Builder().withOperator(windowOperator).build(),
+                        newTopNOptExp);
+
+        OptExpression newProjectOptExp =
+                OptExpression.create(new LogicalProjectOperator.Builder().withOperator(projectOperator).build(),
+                        newWindowOptExp);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java:139]: Don't need create new operator if wasn't change~

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+        if (windowOperator.getWindowCall().size() != 1) {
+            return false;
+        }
+
+        ColumnRefOperator windowCol = Lists.newArrayList(windowOperator.getWindowCall().keySet()).get(0);
+        CallOperator callOperator = windowOperator.getWindowCall().get(windowCol);
+
+        // TODO(hcf) we support dense_rank later
+        if (!FunctionSet.ROW_NUMBER.equalsIgnoreCase(callOperator.getFnName()) &&
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java:64]: Only need equals, all function names are lower-case

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+ * For ranking window functions, such as row_number, rank, dense_rank, if there exists rank related order by clause
+ * and limit clause, then we can add a TopN to filter data in order to reduce the amount of data to be exchanged and sorted
+ * E.g.
+ * select * from (
+ * select *, rank() over (order by v2) as rk from t0
+ * ) sub_t0
+ * order by rk
+ * limit 5;
+ */
+public class PushDownLimitRankingWindowRule extends TransformationRule {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java:37]: would better comment the before and after tree

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java (hunk 1 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java:139]: ok

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java (hunk 3 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java:53]: updated

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java (hunk 1 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java]: updated

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 64 / 67 (95.52%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownLimitRankingWindowRule.java|55|58|94.83%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateRankingWindowRule.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**